### PR TITLE
#905: single core concat (tile layout)

### PR DESCRIPTION
**Pull request by muthutt**
Ref:

I have a proposal for implementing torch.cat() using transposes; we transpose both input matrices to have last dimension to be concat then we reverse the transpose of the output. This works out conceptually in Torch/Python matching output of torch.cat 
```python
tensor([[[[0.8768, 0.0536],
          [0.6373, 0.1284]]]])
tensor([[[[0.4840, 0.7535],
          [0.8339, 0.8506],
          [0.4625, 0.4673]]]])
torch.cat((a,b),2)
tensor([[[[0.8768, 0.0536],
          [0.6373, 0.1284],
          [0.4840, 0.7535],
          [0.8339, 0.8506],
          [0.4625, 0.4673]]]])
proposed solution:

2:45
def tt_cat(a,b,dim):
...     return torch.transpose( torch.cat([ torch.transpose(a,3,dim), torch.transpose(b,3,dim)],3),dim,3)
...
tensor([[[[0.8768, 0.0536],
          [0.6373, 0.1284],
          [0.4840, 0.7535],
          [0.8339, 0.8506],
          [0.4625, 0.4673]]]])
```
underlying logic being concat along dim 3 is cheap (just copy a and b tensors in same order).

- first implementation for tiled only
- second implementation can be for row-major only
-  add documentation
- transpose op scaffolding for missing combinations of NH, NW, and CW respectively
- implement concat through a series of transpose-flatten-concat-reshape-transpose ops

### CODE REVIEW

**Review by tt-aho — CHANGES_REQUESTED** [review]

**Review by ntarafdar — CHANGES_REQUESTED** [review]
A few small changes. thanks Muthu.

**Review by ashayestehmanesh — DISMISSED** [review]

**Review by ntarafdar — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]
lgtm once the final few comments can be addressed.

I think one next step that can come in a separate PR is that we should take the output mem config as an argument, similar to other ops, instead of inferring it from the first input tensor.

**Review by tt-nshanker — APPROVED** [review]
